File: services/auth/src/services/factoryLog.js
/**
 * Writes security events to factory_logs.flg_entries.
 * Uses the `pg` module directly with a parameterized query — no raw SQL strings
 * are constructed from user input; all values flow through pg's parameter binding.
 * Failures are always suppressed: logging must never break the auth service.
 */
const { Pool } = require('pg');

const LOGGING_URL = process.env.LOGGING_DATABASE_URL || '';
let pool = null;

function getPool() {
  if (!pool && LOGGING_URL) {
    pool = new Pool({ connectionString: LOGGING_URL });
  }
  return pool;
}

/**
 * @param {'INFO'|'WARN'|'ERROR'} level
 * @param {string} message
 * @param {object|null} context
 */
async function logEvent(level, message, context = null) {
  const p = getPool();
  if (!p) return;
  try {
    await p.query(
      'INSERT INTO factory_logs.flg_entries (app_label, level, message, context) VALUES ($1, $2, $3, $4)',
      ['auth', level.toUpperCase(), message, context ? JSON.stringify(context) : null]
    );
  } catch {
    // Logging failures must not propagate
  }
}

module.exports = { logEvent };

File: services/export/factory_log.py
"""
Writes operational events to factory_logs.flg_entries.
Uses SQLAlchemy expression language — no raw SQL.
Failures are suppressed: logging must never break the service.
"""
import os

from sqlalchemy import Column, MetaData, String, Table, Text, insert
from sqlalchemy import create_engine
from sqlalchemy.dialects.postgresql import JSONB

_LOGGING_URL: str | None = os.environ.get("LOGGING_DATABASE_URL") or None
_log_engine = None

_metadata = MetaData()
_flg_entries = Table(
    "flg_entries",
    _metadata,
    Column("app_label", String(64), nullable=False),
    Column("level", String(16), nullable=False),
    Column("message", Text, nullable=False),
    Column("context", JSONB, nullable=True),
)


def _get_log_engine():
    global _log_engine
    if _log_engine is None and _LOGGING_URL:
        _log_engine = create_engine(_LOGGING_URL, pool_pre_ping=True)
    return _log_engine


def log_event(level: str, message: str, context: dict | None = None) -> None:
    engine = _get_log_engine()
    if engine is None:
        return
    try:
        with engine.connect() as conn:
            conn.execute(
                insert(_flg_entries).values(
                    app_label="export",
                    level=level.upper(),
                    message=message,
                    context=context,
                )
            )
            conn.commit()
    except Exception:
        pass
